### PR TITLE
Added Sequential module

### PR DESCRIPTION
**Pull request by yushiyangk**
Similar to PyTorch, this is useful for defining repeated sequences of modules (for example in Transformer).

### CODE REVIEW

File: shumai/tensor/register_gradients.ts [hunk 1]
```diff
@@ -131,11 +131,23 @@ const impls = {
   },
   matmul: (grad: Grad) => {
     if (grad.idx === 0) {
-      const yT = (<Tensor>grad.in[1]).T()
-      return grad.grad_in.matmul(yT)
+      const y = <Tensor>grad.in[1]
+      if (grad.grad_in.shape.length === 1 && y.shape.length === 1) {
+        // grad_in and y are 1D column vectors
```
Comment (cryptodeal) [inline shumai/tensor/register_gradients.ts:136]: Appreciate the comments throughout the code here!

File: test/utils.ts [hunk 2]
```diff
@@ -68,8 +68,8 @@ export const isClose = (actual: number | bigint, expected: number | bigint, erro
   if (typeof actual !== typeof expected) return false
 
   if (typeof actual === 'bigint' && typeof expected === 'bigint') {
-    const upper = expected + BigInt(error),
-      lower = expected - BigInt(error)
+    const upper = Number(expected) + error,
```
Comment (cryptodeal) [inline test/utils.ts:71]: The changes here could potentially cause issues if the `expected` value is `typeof BigInt` and a value greater than can be represented as a javascript native Float64.

How would you feel about changing it to:
```ts
export const isClose = (
  actual: number | bigint,
  expected: number | bigint,
  error = 0.001,
  castToNum = false
) => {
  if (typeof actual !== typeof expected) return false

  if (typeof actual === 'bigint' && typeof expected === 'bigint' && !castToNum) {
    const upper = expected + BigInt(error),
      lower = expected - BigInt(error)
    if (actual < lower || actual > upper) {
      return false
    }
    return true
  } else if (typeof actual === 'bigint' && typeof expected === 'bigint' && castToNum) {
    const upper = Number(expected) + error,
      lower = Number(expected) - error
    if (actual < lower || actual > upper) {
      return false
    }
  } else if (typeof actual === 'number' && typeof expected === 'number') {
    const upper = (expected as number) + error,
      lower = expected - error
    if (actual < lower || actual > upper) {
      return false
    }
    return true
  }
}
```

That way we can pass in `castToNum = true` and opt in to this behavior?

Comment (yushiyangk) [reply test/utils.ts:71]: I think this was an unrelated change that got accidentally included. However, the issue was that `BigInt(error)` always gave me an error when `error` is not an integer (which it usually isn't).

Comment (yushiyangk) [reply test/utils.ts:71]: I will probably revert this change for this PR.

This should probably be fixed in a separate PR, but I'm not entirely sure what would be the best way to approach it.

Comment (cryptodeal) [reply test/utils.ts]: Ohhhh, good catch! I had hit that elsewhere in the code, this should fix it:

```ts
export const isClose = (actual: number | bigint, expected: number | bigint, error = 0.001) => {
  if (typeof actual !== typeof expected) return false

  if (typeof actual === 'bigint' && typeof expected === 'bigint') {

    const factor = 1 / error
    const upper = expected * BigInt(factor) + BigInt(error * factor),
      lower = expected * BigInt(factor) - BigInt(error * factor)
    if (actual < lower || actual > upper) {
      return false
    }
    return true
  } else if (typeof actual === 'number' && typeof expected === 'number') {
    const upper = (expected as number) + error,
      lower = expected - error
    if (actual < lower || actual > upper) {
      return false
    }
    return true
  }
}
```

File: shumai/module/sequential.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import { Module } from './module'
+
+import type { Tensor } from '../tensor'
+
+export class Sequential extends Module {
+  private modules: Module[]
+
+  constructor(modules: Module[]) {
```
Comment (bwasti) [inline shumai/module/sequential.ts:8]: PyTorch uses the equivalent of `...modules`, perhaps we should do the same?

File: shumai/module/sequential.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+  get length(): number {
+    return this.modules.length
+  }
+
+  at(index: number): Module {
+    return this.modules.at(index)
+  }
+
+  push(module: Module): number {
```
Comment (bwasti) [inline shumai/module/sequential.ts:21]: curious: why would we want all these list operations?  (I'm not against it)

File: test/utils.ts (hunk 2 above)
Comment (cryptodeal) [reply test/utils.ts]: I've submitted a separate PR with the above fix, so disregard this feedback.

**Comment by bwasti** [comment]
lgtm!

**Pull request merged**